### PR TITLE
Add unique morphisms from/to `Initial` and `Terminal` algebras

### PR DESCRIPTION
**Pull request by jamesmckinna**
Following up on #1906 , but with different location for the constructions.
- [x] `Monoid`
- [x] `Group`
- [x] `NearSemiring`
- [x] `Semiring`
- [x] `Ring`

PR does *not* introduce `Algebra.Morphism.Construct.Zero`, by contrast with `Algebra.Construct.Zero`, because the initial/terminal *morphisms* are different, even while the zero *objects* are identical...

### CODE REVIEW

File: src/Algebra/Morphism/Construct/Terminal.agda [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+private
+  variable
+    a ℓa : Level
+
+
+------------------------------------------------------------------------
+-- The underlying data of the morphism
+
+module 𝕆neMorphism (M : RawMonoid a ℓa) where
```
Comment (JacquesCarette) [inline src/Algebra/Morphism/Construct/Terminal.agda:40]: That module name is perhaps overly cute?

**Review by JacquesCarette — COMMENTED** [review]
Looking good.

File: src/Algebra/Morphism/Construct/Terminal.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Algebra/Morphism/Construct/Terminal.agda:40]: I'd wanted to overload `𝕆ne` from `Algebra.Construct.Terminal`, but resolving the disambiguation was horrible. 
As for 'cute', sure, but when writing `Algebra.Construct.Terminal` and `Algebra.Construct.Initial`, i had been trying to write a *single* internal module definition such that all the relevant definitions could be written as `record { 𝕆ne }` etc. (it nearly worked, but not quite). 

I *do* like blackboard-bold identifiers for 'canonical' constructions, but `\b1` doesn't display in my Emacs... :-(

Comment (jamesmckinna) [reply src/Algebra/Morphism/Construct/Terminal.agda:44]: Alternative: `The{Unique}Morphism`?

**Comment by JacquesCarette** [comment]
I think I liked `OneMorphism` and `ZeroMorphism` (without the cuteness) more than `The`. In some ways, if these files are meant to be used qualified, then just `Morphism` might work, so that they would be `Terminal.Morphism` and `Initial.Morphism`.

**Comment by jamesmckinna** [comment]
> I think I liked `OneMorphism` and `ZeroMorphism` (without the cuteness) more than `The`. In some ways, if these files are meant to be used qualified, then just `Morphism` might work, so that they would be `Terminal.Morphism` and `Initial.Morphism`.

I don't want to make this a hill on which either of us die defending... I'd be happy/ier with plain `Morphism` than `One` or `Zero` (I agree with removing the cuteness), but I liked the rhetorical force (not: cuteness!) of the `The` (`Unique...` would also work?).

**Comment by JacquesCarette** [comment]
`Unique` has a certain charm to it, I agree.

File: src/Algebra/Morphism/Construct/Initial.agda [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Level using (Level)
+
+module Algebra.Morphism.Construct.Initial {c ℓ : Level} where
+
+open import Algebra.Bundles using (Magma)
+open import Algebra.Bundles.Raw using (RawMagma)
+open import Algebra.Construct.Initial {c} {ℓ}
```
Comment (MatthewDaggitt) [inline src/Algebra/Morphism/Construct/Initial.agda:19]: It really confused me where `c` and `l` were being used. Can we move imports that rely on module parameters between the normal inputs and the variable declarations? (Might be good to add to the style guide?)

File: src/Algebra/Morphism/Construct/Initial.agda [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  homo : Homomorphic₂ zero ℤero._∙_ _∙_
+  homo ()
+
+  injective : Injective ℤero._≈_ _≈_ zero
+  injective {x = ()}
+
+------------------------------------------------------------------------
+-- Magma
+
+module _ (M : Magma m ℓm) where
```
Comment (MatthewDaggitt) [inline src/Algebra/Morphism/Construct/Initial.agda:57]: Can't/shouldn't this and all the similar modules in `Terminal` be take raw versions of the bundle as parameters instead?

You could then save a lot of lines (and IMO maybe make it easier to read?) by opening `MagmaMorphisms` etc. at the top of the file without a parameter?

File: src/Algebra/Morphism/Construct/Terminal.agda [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+------------------------------------------------------------------------
+-- Ring
+
+module _ (R : Ring a ℓa) where
+
+  private module R = Ring R
+  open RingMorphisms R.rawRing rawRing
+  open UniqueMorphism R.+-rawMonoid
```
Comment (MatthewDaggitt) [inline src/Algebra/Morphism/Construct/Terminal.agda:141]: Missing empty line?

File: src/Algebra/Morphism/Construct/Terminal.agda [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        ; module RingMorphisms
+        )
+open import Data.Product.Base using (_,_)
+open import Function.Definitions using (StrictlySurjective)
+import Relation.Binary.Morphism.Definitions as Definitions
+open import Relation.Binary.Morphism.Structures
+
+private
+  variable
+    a ℓa : Level
```
Comment (MatthewDaggitt) [inline src/Algebra/Morphism/Construct/Terminal.agda:38]: For better or worse the current standard name is `ℓ` in the library.

Comment (jamesmckinna) [reply src/Algebra/Morphism/Construct/Terminal.agda:38]: Yes, but the trouble is, `c` and `ℓ` are already taken by the top-level parametrisation of the module...

File: src/Algebra/Morphism/Construct/Initial.agda (hunk 3 above)
Comment (jamesmckinna) [reply src/Algebra/Morphism/Construct/Initial.agda:58]: Hmmm... that's an interesting/tricky one. Clearly, for each of `Initial` and `Terminal`, there is the underlying module parameterised on a `Raw` bundle.

For whatever reason (?), I associate `Construct`s as being parametrised on the *bundle*, proxying for the 'name of the category'. But you're quite correct that all the other modules in that directory only parametrise on the `Raw` thing (?), so for consistency's sake, I should revise as you suggest.

Comment (jamesmckinna) [reply src/Algebra/Morphism/Construct/Initial.agda]: Not sure I understand how to save the lines though. Will push the re-parametrisation, and then maybe you could offer me a steer on that?

Comment (MatthewDaggitt) [reply src/Algebra/Morphism/Construct/Initial.agda]: I've rearranged it to get rid of almost 70 lines of boilerplate. Let me know if you think I've lost anything crucial?

**Review by JacquesCarette — APPROVED** [review]

File: src/Algebra/Morphism/Construct/Initial.agda (hunk 3 above)
Comment (jamesmckinna) [reply src/Algebra/Morphism/Construct/Initial.agda]: Looks good, modulo it now being harder for me to read, but that's a common problem! 

Your desire for brevity contrasts with what I regard as mine for (greater!?) clarity, but it does seem to simply be a contrast in styles. 

Happy to let yours carry the day. Not sure how well I can internalise it though! ;-)

**Comment by jamesmckinna** [comment]
Happy to merge, and will try to abstain in future from my inclination towards boilerplate/bureaucracy...

**Pull request merged**